### ISSUE TITLE
Convert nav and random into UPM-compatible packages

### ISSUE THREAD

**Issue by reeseschultz** [reeseschultz/ReeseUnityDemos#9]
The process of finding and using my scripts and demos would be vastly improved if they were available in UPM. Ideally the entire `Reese.Nav` namespace and respective file hierarchy could be its own package without need to migrate to another repository, while still having a bigger package for everything else that includes nav. I'll look into this.

**Comment by reeseschultz** [comment]
Ugh. This seems like a massive pain. There's a bunch of UPM-related tools and registries doing things slightly differently. Seems like nobody's pinned down a consistent, simple and singular process. Cool.

Regardless, I'll likely need to set up semver/git tags, which means figuring out what "version" the project already is, or the nav part at least. Initializing to `1.0.0` makes sense, insofar as versioning makes sense here anyway. (Versioning projects like this for other developers never made much sense to me, considering there's a commit hash telling us all we need to know.)

If anything gets published to UPM there should probably be tests, which isn't a complaint so much as it is an exasperated sigh. I'm not keen on writing tests for something that's changing at a million miles per hour. I figure after #8 is tackled, I'll be comfortable refactoring the job code into unit-test-friendly functions, since at that point I doubt expectations of them will change much.

Some automation will have to be introduced as well, not necessarily to build anything, but to at least handle publishing, which can't be too bad.

**tl;dr** I'm not too worried about this issue right now, since there's other stuff I need to do first.

**Comment by reeseschultz** [comment]
Despite an overcomplicated ecosystem, UPM itself doesn't care about semver, just the [commit hashes like I do](https://docs.unity3d.com/Manual/upm-git.html). Tests are recommended, but not needed.

Excellent, I may get started on this soon since I'm blocked by a Unity bug on some other issues.

**Comment by reeseschultz** [comment]
Okay, for now I can semver my packages at `0.0.0` (or `0.0.1`). When people add them as a Git packages in Unity, it'll automatically lock at `HEAD` on the current commit hash. I don't know if it can find updates that way, but the worst-case scenario is that users could just remove and add the package they want back.

*Supposedly* I can create sub-packages for `Reese.Nav` and `Reese.Random`. I'm *pretty* sure I can also create a root package for the entire repo that references those subpackages as "embedded" packages. I'll create a different branch to test this stuff. One thing I'm hoping is that the dependencies for the subpackages are auto-detected from `Packages/manifest.json`. If not, that would be super annoying to have to define duplicate data in the json files....

**Comment by reeseschultz** [comment]
There's initial UPM-ification in the [upm-git](https://github.com/reeseschultz/ReeseUnityDemos/tree/upm-git) branch. That was seriously not fun, especially trying to figure out that I needed `Entities.Hybrid` and `Transforms.Hybrid` in some `asmdef` files...

I'll continue working in that branch, because there's more to do (changing access modifiers due to using separate assemblies now; testing Git imports into another project; changing the nav user guide links to reflect new file hierarchy; etc), then I'll squash the commits and merge them into master.

**Comment by reeseschultz** [comment]
I have no way of testing the import into another project from another branch. Awesome. What I'm wondering about is how having multiple packages will work. I have a root package, and then two other packages. It may be that the Git import for UPM won't be sufficient if it can't tell there are actually multiple packages, meaning I *would* have to add actual semver and then come up with a publishing process for the subpackages.

**Comment by reeseschultz** [comment]
According to [this](https://forum.unity.com/threads/git-support-on-package-manager.573673/page-5), looks like Unity's working on support for handling multiple Git-based packages in one repo and it should be available in `2020.1`.

**Comment by reeseschultz** [comment]
Closing because of 87f14d750d5da93e860708988c93120866432feb.
